dataservice/api/common/pagination.py

```diff
@@ -0,0 +1,86 @@
+from flask import request, current_app
+from functools import wraps
+from dateutil import parser
+from datetime import datetime
+
+
+def paginated(f):
+
+    @wraps(f)
+    def paginated_wrapper(*args, **kwargs):
+        def_limit = current_app.config['DEFAULT_PAGE_LIMIT']
+        max_limit = current_app.config['MAX_PAGE_LIMIT']
+        limit = min(request.args.get('limit', def_limit, type=int), max_limit)
+        after = request.args.get('after', None)
+
+        if type(after) is str:
+            # Parser won't recognize the timestamp with fractional seconds
+            if after.replace('.', '').isdigit():
+                after = float(after)
+                after = datetime.fromtimestamp(after)
+            else:
+                try:
+                    after = parser.parse(after)
+                # Parser couldn't derive a datetime from the string
+                except ValueError:
+                    after = None
+
+        # Default to the unix epoch
+        if after is None:
+            after = datetime.fromtimestamp(0)
+
+        return f(*args, **kwargs, after=after, limit=limit)
+
+    return paginated_wrapper
+
+
+class Pagination(object):
+    """
+    Object to help paginate through endpoints using the created_at field
+    """
+
+    def __init__(self, query, after, limit):
+        assert type(after) is datetime
+        self.query = query
+        self.after = after
+        self.limit = limit
+        self.total = query.count()
+        # Assumes that we only provide queries for one entity
+        # This is safe as pagination only accesses one entity at a time
+        model = query._entities[0].mapper.entity
+        assert hasattr(model, 'created_at')
+        self.items = (query.order_by(model.created_at.asc())
+                           .filter(model.created_at > after)
+                           .limit(limit).all())
+
+    @property
+    def prev_num(self):
+        """ Returns the timestamp of the first item """
+        if len(self.items) > 0:
+            return self._to_timestamp(self.items[0].created_at)
+        return datetime.fromtimestamp(0)
+
+    @property
+    def curr_num(self):
+        """
+        Returns the timestamp of the first item minus some epsilon
+        so that the current timestamp will be included in the range
+        """
+        if len(self.items) > 0:
+            return self._to_timestamp(self.items[0].created_at) - 1/1000
+        return datetime.fromtimestamp(0)
```

Comment:
Do we want to default to unix epoch here? Should it just return the request date `after` with no results? Or maybe just default to the first page.

Right now when you make a request that should return no results: 
`localhost:5000/participants?after=2018-02-28`

I get back:
```
{
    "_links": {
        "self": "/participants?after=1969-12-31+19%3A00%3A00"
    },
    "_status": {
        "code": 200,
        "message": "success"
    },
    "limit": 10,
    "results": [],
    "total": 1203
}
``` 
I think this needs to change to 0 results but with the self link set to the request `after` date in the original request or first page of results by default with the `after` date kept as is (unix epoch)

Reply:
Yeah, doesn't make sense to return epoch. We would expect some number of results then. Likewise, it doesn't make sense to return the first page, either.

I'm going to default to whatever the requested url was using request.full_path, and default to the current timestamp for the this property.